### PR TITLE
File System API (Final On-Hold) See Comments

### LINKED ISSUE #329: Add simple filesystem backend
**- Do you want to request a *feature* or report a *bug*?**

New feature.

**- What is the current behavior?**

Currently there is no backend which uses the filesystem. This would be useful for local work and testing.

**- What is the expected behavior?**

There should be a backend available which calls a simple API that allows reading and writing from the filesystem. Netlify CMS should allow you to spin up a local filesystem-backed instance via an `npm` task.

### PR DESCRIPTION
**Pull request by talves**
***Note:*** This PR is on hold, because the solution will be moved to an outside proposed mono-repo. See [this repo][1] for a working solution.

**- Summary**

Create an end point file system api middleware for the express server (webpack-dev-server) to be able to create a backend api in the CMS to update the example site locally.

1. Checkout the branch `yarn` and `yarn start`
2. The devServer will list two paths. The root path to the project location and the path to the site directory. These paths should be correct.

Example on windows:
```
Root path is D:\Git\Netlify\netlify-cms-origin\
Site path is D:\Git\Netlify\netlify-cms-origin\example
```

Closes #329
Closes #710 
Closes #706 

**- Test plan**

- [ ] Needs a test on POSIX systems
- [x] Needs a test on MAC
- [x] Windows has been tested but another look would be good

**- Description for the changelog**
Features:

- Able to run against the `/api` path at two endpoints `/api/file/:id` and `/api/files/:path`
- The api will later be able to run as a stand alone middleware plugin for a local express server with no changes (written)

@Benaiah  @erquhart I will explain as much as I can on this PR, so you can test quicker.

I have created a new example structure at the root of `example` the way Shawn likes it 😄 It has the same paths and folders as the orphan branch `cms-test-backend`.

## Structure for deploy:
I created a `config.yml` file in the assets folder to use when a deploy preview kicks off from now on that will point to the orphan branch. All someone has to do on a test branch that needs a different layout is put their copy of the config in there and it will now get copied over to the root `example/config.yml` when the PR deploy preview is created.

## API
The new ***backend*** API is called `file-system` in the cms. It targets the endpoints in the API middleware in the `scripts/fs` folder. When a webpack dev server is started, the api routes are added to the server.

## Enhancements (updated: 11-8-2017)

- defaults to file-system during development (should only commit changes when default changes)
- copies `production/config.yml` and `production/index.html` on a `production` deploy (these should not change except on design changes)
- copies `preview/config.yml` on a `branch-deploy` and `deploy-preview` (these are ok to change for your branch targets)

## Notes:
I think all the backend API's should be plugins at some point for the cms, so we are not carrying around unnecessary code.

The failures on read or write do make the cms fail elegantly. Any errors will show up in the inspect of the page as 500 errors.

Hope this all makes sense and code like you are on :fire:

I will close #781 due to conflicts on my rebase of the branch.

[1]: https://github.com/talves/netlify-cms-backend-fs

### CODE REVIEW

File: webpack.dev.js [hunk 1]
```diff
@@ -61,6 +61,6 @@ module.exports = merge.smart(require('./webpack.base.js'), {
     historyApiFallback: true,
     disableHostCheck: true,
     headers: {"Access-Control-Allow-Origin": "*"},
-    setup: fileSystemAPIPlugin,
+    before: fileSystemAPIPlugin,
```
Comment (talves) [inline webpack.dev.js]: @tech4him1 good catch, I ran this for the last 4 days and did not notice the warning 👍

File: example/assets/preview/config.yml [hunk 2]
```diff
@@ -0,0 +1,166 @@
+backend:
+  name: github
+  repo: netlify/netlify-cms
+  branch: cms-test-backend
```
Comment (tech4him1) [inline example/assets/preview/config.yml]: If we use the CMS repo on the GitHub backend for deploy previews, contributors that do not have write permissions will not be able to use our deploy previews. Right?

Comment (talves) [reply example/assets/preview/config.yml]: True. That branch is an orphan branch we setup to test the backend, so in most cases we would just have to give those access rights to contributors that need them.

Do you think this is an issue? Maybe hard to maintain?

***On another note:*** We could always lock files in the paths for production and this one to make sure there is not an overwrite. Not sure we want that, but an idea.

Comment (talves) [reply example/assets/preview/config.yml]: @erquhart can you let me know if what Caleb is asking here is an issue, so we can address?
The solution is fine either way, just want to make sure we don't introduce any issues on deploy previews.

**Comment by talves** [comment]
There has been a rebase to version 0.7.1 here, so make sure when pulling this branch, you pull a clean checkout from origin.

File: src/backends/file-system/API.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        return fileTree;
+      }
+    });
+  }
+
+  deleteFile(path, message, options={}) {
+    const fileURL = `/file/${ path }`;
+    return this.request(fileURL, {
+      method: "DELETE",
+      params: { },
```
Comment (tech4him1) [inline src/backends/file-system/API.js]: These `params: { },` should be necessary.

File: src/backends/test-repo/implementation.js [hunk 4]
```diff
@@ -26,7 +26,7 @@ function nameFromEmail(email) {
 export default class TestRepo {
   constructor(config) {
     this.config = config;
-    this.assets = [];
+    this.assets = window.repoFiles.assets || [];
```
Comment (tech4him1) [inline src/backends/test-repo/implementation.js]: Is this line related to this PR at all?

Comment (talves) [reply src/backends/test-repo/implementation.js]: https://github.com/netlify/netlify-cms/pull/786/files#diff-1634bb1b7f4da7ca72cddcb74fef74b7R12

Allows for the loading of the assets that are truly in the `assets/images` folder. The way they would show up if you were truly hitting a real api.

File: src/backends/file-system/API.js (hunk 3 above)
Comment (talves) [reply src/backends/file-system/API.js]: right now there is nothing being done with them to the file-system, but maybe later in the case we want to do logging etc in next phase.

File: src/backends/test-repo/implementation.js (hunk 4 above)
Comment (tech4him1) [reply src/backends/test-repo/implementation.js]: Awesome!

**Comment by jrop** [comment]
I have just tested on macOS and this seems to be in working order.  I for one cannot wait for this to land.

**Comment by erquhart** [comment]
Awesome, thanks for testing it out @jrop! We definitely want to get this in soon.

**Comment by jrop** [comment]
I also fired this up in my Ubuntu VM and creating a post worked fine.  I don't know what kind of testing you guys are looking for, though.

**Comment by talves** [comment]
Thanks @jrop 
I was looking to make sure paths added correctly and additions and deletions worked with the commands on those systems. I covered as much as I could to make sure the commands were compatible, so until it goes into the wild, that is the best we can do for now.  I will sync it with the latest version in a while and update the rebase version up top, so we can test against the current version. 😄

**Comment by AstroJetson** [comment]
@talves -  Do you have a guess when this will land?  I'm very interested in using it on a locally hosted server. 

Thanks!

**Comment by talves** [comment]
@AstroJetson I am only a contributor, so if it is accepted @erquhart will have the details on when it will land. I believe it will get looked at after 1.0 is my best guess.

**Comment by erquhart** [comment]
@talves I'm pretty beat from the 1.0 rush, but I'm planning onto digging into this next week. From what I've seen it looks like a _massive_ improvement over the in-memory repo.

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit 5ed7257e99ddf5d0287f188796384476a6cad3cf

https://deploy-preview-786--netlify-cms-www.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit bb40158ff9465420e6a0d914e4df1011fbaff6c1

https://deploy-preview-786--cms-demo.netlify.com

**Comment by erquhart** [comment]
@talves you [mentioned](https://github.com/netlify/netlify-cms/pull/1029#issuecomment-359178101) discussing use of this as an external backend - #1011 has landed, so it's possible. Thoughts?

cc/ @tech4him1 @Benaiah

**Comment by talves** [comment]
There was a lot of discussion on bringing the current internal backends out "into the wild" between me and @tech4him1. There are a lot of considerations and we might have thought it would be better to keep the current backends in place until we hash out a lot of these considerations.

## Things to consider:

- Dependencies
- API methods required
- Documentation
- Maintenance

These are all obvious, but really should be discussed to keep the CMS focused (github)

I have the localhost already completed and will publish it soon (very soon). This alone will help us to spark conversation on if we want to keep localhost fs internal with github, but make others external.  Let the discussion begin. 😁

**Comment by talves** [comment]
[Here is the backend repository][1] for the file-system in this PR

Checkout the `cms-server` branch for a stand alone server example. I plan to merge it to master. User master if it has been merged. If someone wants to set it up in a project ping me in gitter for help.

Eventually this repository will be put into the cms as a mono-repo. I will keep this PR open for discussion of that repository since it will be replacing this PR.

cc/ @tech4him1 @erquhart @Benaiah 

[1]: https://github.com/talves/netlify-cms-backend-fs

**Comment by talves** [comment]
Should we Close ths PR or Leave it (mark on-hold) for reference until we have a viable mono-repo for the complete solution?

**Comment by bjrn** [comment]
@talves if I'd like to test the local backend for development, am I right to assume that I go to the aforementioned repo and try it from there? Will be using it with Gatsby … I guess there's no way of just setting the backend option in config.yml just yet?

**Comment by talves** [comment]
@bjrn yes, you can easily clone/fork [my repo][2] and run a build. You can always hit me up in [Gitter][1] if you need help setting it up. The readme instructions has most of what you need to wire it into your project. 

Eventually this repo will be a mono-repo inside the CMS and will be accessible through npm, but is fully functional and tested for local development to use now if you want in a project.

[1]: https://gitter.im/netlify/NetlifyCMS
[2]: https://github.com/talves/netlify-cms-backend-fs

**Comment by erquhart** [comment]
Closing as stale. @talves deleting branch as the it's backed up on your fork.

**Comment by jrop** [comment]
@erquhart @talves I see that this got closed.  What is the state of a File-System backend going forward for 2.0?  Will that likely be a community-driven module?

**Comment by talves** [comment]
@erquhart No need to back up. It has it's own [repo](https://github.com/talves/netlify-cms-backend-fs)(s) in my account.

@jrop Exists for v1.x. See [this example repo](https://github.com/talves/netlify-cms-react-example/tree/master/src/components/NetlifyCMS/components/FileSystemBackend). I will more than likely move it for v2.x when I get some time. You could always take a stab at it.

Not sure if we will keep it as a community driven backend or move it into the repo, because it needs so much extra setup on top of a project to spin up the File-System API.

**Comment by cjbrooks12** [comment]
Would it make sense to maintain the CMS backend separately from the node server? The backend could then be more of a generic HTTP backend, and any service that implements the REST endpoints can then be dropped in. Separately, the node filesystem server just implements the rest interface.

**Comment by cjbrooks12** [comment]
As motivation for this idea, I've been using the CMS backend portion of this for a while now in [Orchid](https://orchid.netlify.com/plugins/Orchid NetlifyCMS). Orchid has a development server that implements the api needed by the backend, and so local development goes through the Orchid server, rather than the node server.

**Comment by talves** [comment]
> Would it make sense to maintain the CMS backend separately from the node server? The backend could then be more of a generic HTTP backend, and any service that implements the REST endpoints can then be dropped in. Separately, the node filesystem server just implements the rest interface.

@cjbrooks12 these endpoints are not part of the file-system backend, they just happen to be setup in a server using node in the example repo. You could use any server as long as you setup the same endpoints. Are you suggesting something different? Maybe I am misreading your suggestion.

**Comment by cjbrooks12** [comment]
Right, the endpoints and their formats are just what the Node server happens to be spitting out, but if we invert how we think of this, the "filesystem backend" is just a backend that hits a REST API at a given URL, and the actual filesystem portion just happens to implement the correct API endpoints with data from the local filesystem. 

But because the CMS backend is just making HTTP calls, any server that produces the same response can be used and the CMS will work just fine, which is how I've already been using it for a while. I've tossed aside the actual "filesystem" portion of your fork, the node server, and replaced it with Orchid's own development server. Orchid then re-implements (in Kotlin) the code necessary to transform its own concept of pages on disk into the JSON response that the CMS backend is expecting. 

Conceivably, if the actual format that the CMS backend reads were to be formalized and documented, other tools could create ways to implement those endpoints in their own development servers. Think, a Gatsby plugin that allows you to run `gatsby develop` and you'd be able to point the backend at `http://localhost:8000/netlify-cms` and have it magically work as expected, without the need of having a dedicated `gatsby-server` CMS backend, and without needing to start an external service to serve the filesystem contents to Netlify CMS.

**Comment by jrop** [comment]
I second the idea to have a backend creation "spec" or "guide".  I would potentially like to create my own backends for Netlify CMS.
